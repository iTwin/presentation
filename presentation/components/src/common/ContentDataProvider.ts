/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { KeySet, PageOptions, SelectionInfo, ContentRequestOptions } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import * as content from "@bentley/presentation-common";

/**
 * Properties for invalidating content cache.
 */
export interface CacheInvalidationProps {
  /**
   * Invalidate content descriptor. Should be set when invalidating
   * after changing anything that affects how the descriptor is built:
   * `keys`, `selectionInfo`, `connection`, `rulesetId`.
   */
  descriptor?: boolean;

  /**
   * Invalidate configured content descriptor. Should be set when
   * invalidating something that affects how descriptor is configured
   * in the `configureContentDescriptor` callback, e.g. hidden fields,
   * sorting, filtering, etc.
   */
  descriptorConfiguration?: boolean;

  /**
   * Invalidate cached content size. Should be set after changing anything
   * that may affect content size. Generally, it should always be set when
   * the `descriptor` flag is set. Additionally, it should also be set after
   * setting `filterExpression` or similar descriptor properties.
   */
  size?: boolean;

  /**
   * Invalidate cached content. Should be set after changing anything that may
   * affect content. Generally, it should always be set when the `descriptor`
   * flag is set. Additionally, it should also be set after setting `sortingField`,
   * `sortDirection`, `filterExpression` and similar fields.
   */
  content?: boolean;
}
namespace CacheInvalidationProps {
  /**
   * Create CacheInvalidationProps to fully invalidate all caches.
   */
  export const full = (): CacheInvalidationProps => ({ descriptor: true, descriptorConfiguration: true, size: true, content: true });
}

/**
 * Base class for all presentation-driven content providers.
 */
export default abstract class ContentDataProvider {
  private _connection: IModelConnection;
  private _rulesetId: string;
  private _displayType: string;
  private _keys: Readonly<KeySet>;
  private _selectionInfo?: Readonly<SelectionInfo>;

  /**
   * Constructor.
   * @param connection IModel to pull data from.
   * @param rulesetId Id of the ruleset to use when requesting content.
   * @param displayType The content display type which this provider is going to
   * load data for.
   */
  constructor(connection: IModelConnection, rulesetId: string, displayType: string) {
    this._rulesetId = rulesetId;
    this._displayType = displayType;
    this._connection = connection;
    this._keys = new KeySet();
    this.invalidateCache(CacheInvalidationProps.full());
  }

  /** Display type used to format content */
  public get displayType(): string { return this._displayType; }

  /** IModel to pull data from */
  public get connection(): IModelConnection { return this._connection; }
  public set connection(connection: IModelConnection) {
    if (this._connection === connection)
      return;
    this._connection = connection;
    this.invalidateCache(CacheInvalidationProps.full());
  }

  /** Id of the ruleset to use when requesting content */
  public get rulesetId(): string { return this._rulesetId; }
  public set rulesetId(value: string) {
    if (this._rulesetId === value)
      return;
    this._rulesetId = value;
    this.invalidateCache(CacheInvalidationProps.full());
  }

  /** Keys defining what to request content for */
  public get keys() { return this._keys; }
  public set keys(keys: Readonly<KeySet>) {
    this._keys = keys;
    this.invalidateCache(CacheInvalidationProps.full());
  }

  /** Information about selection event that results in content change */
  public get selectionInfo() { return this._selectionInfo; }
  public set selectionInfo(info: Readonly<SelectionInfo> | undefined) {
    if (this._selectionInfo === info)
      return;
    this._selectionInfo = info;
    this.invalidateCache(CacheInvalidationProps.full());
  }

  /**
   * Invalidates cached content.
   */
  protected invalidateCache(props: CacheInvalidationProps): void {
    if (props.descriptor && this.getDefaultContentDescriptor)
      this.getDefaultContentDescriptor.cache.clear!();
    if (props.descriptorConfiguration && this.getContentDescriptor)
      this.getContentDescriptor.cache.clear!();
    if (props.size && this.getContentSetSize)
      this.getContentSetSize.cache.clear!();
    if (props.content && this.getContent)
      this.getContent.cache.clear!();
  }

  private createRequestOptions(): ContentRequestOptions<IModelConnection> {
    return {
      imodel: this._connection,
      rulesetId: this._rulesetId,
    };
  }

  /**
   * Called to configure the content descriptor. This is the place where concrete
   * provider implementations can control things like sorting, filtering, hiding fields, etc.
   *
   * The default method implementation takes care of hiding properties. Subclasses
   * should call the base class method to not lose this functionality.
   */
  protected configureContentDescriptor(descriptor: Readonly<content.Descriptor>): content.Descriptor {
    const fields = descriptor.fields.slice();
    const fieldsCount = fields.length;
    for (let i = fieldsCount - 1; i >= 0; --i) {
      const field = fields[i];
      if (this.shouldExcludeFromDescriptor(field))
        fields.splice(i, 1);
    }
    const customDescriptor = Object.create(content.Descriptor.prototype);
    return Object.assign(customDescriptor, descriptor, content.Descriptor, {
      fields,
    });
  }

  /** Called to check whether the field should be excluded from the descriptor. */
  protected shouldExcludeFromDescriptor(field: content.Field): boolean { return this.isFieldHidden(field); }

  /** Called to check whether the field should be hidden. */
  protected isFieldHidden(_field: content.Field): boolean { return false; }

  // tslint:disable-next-line:naming-convention
  private getDefaultContentDescriptor = _.memoize(async (): Promise<Readonly<content.Descriptor> | undefined> => {
    return await Presentation.presentation.getContentDescriptor(this.createRequestOptions(),
      this._displayType, this.keys, this.selectionInfo);
  });

  /**
   * Get the content descriptor.
   */
  protected getContentDescriptor = _.memoize(async (): Promise<Readonly<content.Descriptor> | undefined> => {
    const descriptor = await this.getDefaultContentDescriptor();
    if (!descriptor)
      return undefined;
    return this.configureContentDescriptor(descriptor);
  });

  /**
   * Get the number of content records.
   */
  protected getContentSetSize = _.memoize(async (): Promise<number> => {
    const descriptor = await this.getContentDescriptor();
    if (!descriptor)
      return 0;
    return await Presentation.presentation.getContentSetSize(this.createRequestOptions(),
      descriptor, this.keys);
  });

  /**
   * Get the content.
   * @param pageOptions Paging options.
   */
  protected getContent = _.memoize(async (pageOptions?: PageOptions): Promise<Readonly<content.Content> | undefined> => {
    const descriptor = await this.getContentDescriptor();
    if (!descriptor)
      return undefined;
    return await Presentation.presentation.getContent({ ...this.createRequestOptions(), paging: pageOptions },
      descriptor, this.keys);
  }, createKeyForPageOptions);
}

const createKeyForPageOptions = (pageOptions?: PageOptions) => {
  if (!pageOptions)
    return "0/0";
  return `${(pageOptions.start) ? pageOptions.start : 0}/${(pageOptions.size) ? pageOptions.size : 0}`;
};
