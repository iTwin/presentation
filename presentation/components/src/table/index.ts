/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Table */

export { default as PresentationTableDataProvider } from "./DataProvider";
export { default as withUnifiedSelection } from "./WithUnifiedSelection";
