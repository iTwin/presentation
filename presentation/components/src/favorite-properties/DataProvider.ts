/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @module FavoriteProperties */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { CategoryDescription } from "@bentley/presentation-common";
import { IElementPropertyDataProvider, PropertyData } from "@bentley/ui-components";
import { Presentation, getScopeId } from "@bentley/presentation-frontend";
import { translate } from "../common/Utils";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider";

const favoritesCategoryName = "Favorite";
let favoritesCategoryPromise: Promise<CategoryDescription> | undefined;

/** @internal */
export const getFavoritesCategory = async (): Promise<CategoryDescription> => {
  if (!favoritesCategoryPromise) {
    favoritesCategoryPromise = Promise.all([
      translate("categories.favorite.label"),
      translate("categories.favorite.description"),
    ]).then(([label, description]): CategoryDescription => ({
      name: favoritesCategoryName,
      label,
      description,
      priority: Number.MAX_VALUE,
      expand: true,
    }));
  }
  return favoritesCategoryPromise;
};

/** @beta */
export interface FavoritePropertiesDataProviderProps {
  /** @internal */
  propertyDataProviderFactory?: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @beta
 */
export class FavoritePropertiesDataProvider implements IElementPropertyDataProvider {

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: null, undefined, "" (empty string)
   * - For *array* fields: [] (empty array)
   * - For *struct* fields: {} (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /**
   * ID of a custom ruleset to use when requesting properties.
   */
  public customRulesetId: string | undefined;

  private _propertyDataProviderFactory: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._propertyDataProviderFactory = props && props.propertyDataProviderFactory ?
      props.propertyDataProviderFactory : /* istanbul ignore next */
      (imodel: IModelConnection, rulesetId?: string) => new PresentationPropertyDataProvider(imodel, rulesetId);
  }

  /**
   * Returns PropertyData for the specified element.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementId: string): Promise<PropertyData> {
    const key = await Presentation.selection.scopes.computeSelection(
      imodel,
      elementId,
      getScopeId(Presentation.selection.scopes.activeScope));

    const propertyDataProvider = this._propertyDataProviderFactory(imodel, this.customRulesetId);
    propertyDataProvider.keys = key;
    propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
    propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
    const propertyData = await propertyDataProvider.getData();

    // leave only favorite properties
    const favoritesCategory = await getFavoritesCategory();
    propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
    propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name) ?
      { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] } : {};

    return propertyData;
  }
}
