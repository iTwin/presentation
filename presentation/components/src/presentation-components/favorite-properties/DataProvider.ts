/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module FavoriteProperties
 */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { CategoryDescription } from "@bentley/presentation-common";
import { IElementPropertyDataProvider, PropertyData } from "@bentley/ui-components";
import { Presentation, getScopeId } from "@bentley/presentation-frontend";
import { translate } from "../common/Utils";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider";

/** @internal */
export const FAVORITES_CATEGORY_NAME = "Favorite";

/** @internal */
export const getFavoritesCategory = (): CategoryDescription => {
  return {
    name: FAVORITES_CATEGORY_NAME,
    label: translate("categories.favorite.label"),
    description: translate("categories.favorite.description"),
    priority: Number.MAX_VALUE,
    expand: true,
  };
};

/** @beta */
export interface FavoritePropertiesDataProviderProps {
  /** @internal */
  propertyDataProviderFactory?: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @beta
 */
export class FavoritePropertiesDataProvider implements IElementPropertyDataProvider {

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: null, undefined, "" (empty string)
   * - For *array* fields: [] (empty array)
   * - For *struct* fields: {} (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /**
   * ID of a custom ruleset to use when requesting properties.
   */
  public customRulesetId: string | undefined;

  private _propertyDataProviderFactory: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._propertyDataProviderFactory = props && props.propertyDataProviderFactory ?
      props.propertyDataProviderFactory : /* istanbul ignore next */
      (imodel: IModelConnection, rulesetId?: string) => new PresentationPropertyDataProvider(imodel, rulesetId);
  }

  /**
   * Returns PropertyData for the specified element.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementId: string): Promise<PropertyData> {
    const key = await Presentation.selection.scopes.computeSelection(
      imodel,
      elementId,
      getScopeId(Presentation.selection.scopes.activeScope));

    const propertyDataProvider = this._propertyDataProviderFactory(imodel, this.customRulesetId);
    propertyDataProvider.keys = key;
    propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
    propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
    const propertyData = await propertyDataProvider.getData();

    // leave only favorite properties
    const favoritesCategory = getFavoritesCategory();
    propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
    propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name) ?
      { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] } : {};

    return propertyData;
  }
}
