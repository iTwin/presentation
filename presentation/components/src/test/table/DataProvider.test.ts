/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/* tslint:disable:no-direct-imports */

import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import { expect } from "chai";
import * as sinon from "sinon";
import * as faker from "faker";
import * as moq from "@bentley/presentation-common/lib/test/_helpers/Mocks";
import { createRandomDescriptor, createRandomECInstanceKey } from "@bentley/presentation-common/lib/test/_helpers/random";
import { PromiseContainer } from "@bentley/presentation-common/lib/test/_helpers/Promises";
import { SortDirection } from "@bentley/ui-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import {
  PresentationError, ValuesDictionary, Content,
  DefaultContentDisplayTypes, Descriptor, Item,
  SortDirection as ContentSortDirection,
} from "@bentley/presentation-common";
import { Presentation, PresentationManager } from "@bentley/presentation-frontend";
import { PresentationTableDataProvider } from "../../table/DataProvider";
import { CacheInvalidationProps } from "../../common/ContentDataProvider";

/**
 * This is just a helper class to provide public access to
 * protected methods of PresentationTableDataProvider
 */
class Provider extends PresentationTableDataProvider {
  public invalidateCache(props: CacheInvalidationProps) { super.invalidateCache(props); }
  public configureContentDescriptor(descriptor: Descriptor) { return super.configureContentDescriptor(descriptor); }
}

interface MemoizedCacheSpies {
  getColumns: sinon.SinonSpy;
}

describe("TableDataProvider", () => {

  let rulesetId: string;
  let provider: Provider;
  let memoizedCacheSpies: MemoizedCacheSpies;
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();
  const imodelMock = moq.Mock.ofType<IModelConnection>();
  before(() => {
    rulesetId = faker.random.word();
    Presentation.presentation = presentationManagerMock.object;
  });
  beforeEach(() => {
    presentationManagerMock.reset();
    provider = new Provider({ imodel: imodelMock.object, ruleset: rulesetId });
    memoizedCacheSpies = {
      getColumns: sinon.spy(provider.getColumns.cache, "clear"),
    };
  });

  const resetMemoizedCacheSpies = () => {
    memoizedCacheSpies.getColumns.resetHistory();
  };

  const createEmptyContentItem = (): Item => {
    return new Item([createRandomECInstanceKey()], faker.random.words(),
      "", undefined, {}, {}, []);
  };
  const createContent = (recordsCount: number, itemsGenerator: () => Item = createEmptyContentItem): Content => {
    const descriptor = createRandomDescriptor();
    const records = new Array<Item>();
    while (recordsCount--) {
      records.push(itemsGenerator());
    }
    return {
      descriptor,
      contentSet: records,
    };
  };
  const createSingleRecordContent = (itemsGenerator?: () => Item) => createContent(1, itemsGenerator);

  describe("constructor", () => {

    it("sets display type to GRID", () => {
      expect(provider.displayType).to.eq(DefaultContentDisplayTypes.GRID);
    });

    it("sets default sorting properties", () => {
      expect(provider.sortColumnKey).to.be.undefined;
      expect(provider.sortDirection).to.eq(SortDirection.NoSort);
    });

  });

  describe("invalidateCache", () => {

    it("resets filtering, sorting, memoized columns and raises onColumnsChanged event when 'descriptor' flag is set", async () => {
      const onColumnsChangedSpy = sinon.spy(provider.onColumnsChanged, "raiseEvent");
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => createRandomDescriptor());

      provider.filterExpression = faker.random.words();
      await provider.sort(0, SortDirection.Descending);

      resetMemoizedCacheSpies();
      onColumnsChangedSpy.resetHistory();

      provider.invalidateCache({ descriptor: true });

      expect(memoizedCacheSpies.getColumns).to.be.calledOnce;
      expect(onColumnsChangedSpy).to.be.calledOnce;
    });

    it("resets memoized columns and raises onColumnsChanged event when 'descriptorConfiguration' flag is set", () => {
      const onColumnsChangedSpy = sinon.spy(provider.onColumnsChanged, "raiseEvent");

      provider.invalidateCache({ descriptorConfiguration: true });

      expect(memoizedCacheSpies.getColumns).to.be.calledOnce;
      expect(onColumnsChangedSpy).to.be.calledOnce;
    });

    it("resets cached pages and raises onRowsChanged event when 'size' flag is set", async () => {
      const onRowsChangedSpy = sinon.spy(provider.onRowsChanged, "raiseEvent");

      const getContentMock = moq.Mock.ofInstance((provider as any).getContent);
      getContentMock.setup((x) => x(moq.It.isAny())).returns(async () => createSingleRecordContent());
      (provider as any).getContent = getContentMock.object;

      expect(await provider.getRow(0)).to.not.be.undefined;
      expect(provider.getLoadedRow(0)).to.not.be.undefined;

      provider.invalidateCache({ size: true });

      expect(provider.getLoadedRow(0)).to.be.undefined;
      expect(onRowsChangedSpy).to.be.calledOnce;
    });

    it("resets cached pages and raises onRowsChanged event when 'content' flag is set", async () => {
      const onRowsChangedSpy = sinon.spy(provider.onRowsChanged, "raiseEvent");

      const getContentMock = moq.Mock.ofInstance((provider as any).getContent);
      getContentMock.setup((x) => x(moq.It.isAny())).returns(async () => createSingleRecordContent());
      getContentMock.object.cache = { clear: () => { } };
      (provider as any).getContent = getContentMock.object;

      expect(await provider.getRow(0)).to.not.be.undefined;
      expect(provider.getLoadedRow(0)).to.not.be.undefined;

      provider.invalidateCache({ content: true });

      expect(provider.getLoadedRow(0)).to.be.undefined;
      expect(onRowsChangedSpy).to.be.calledOnce;
    });

  });

  describe("configureContentDescriptor", () => {

    it("sets sorting properties", async () => {
      const source = createRandomDescriptor();
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);

      await provider.sort(0, SortDirection.Descending);
      let result = provider.configureContentDescriptor(source);
      expect(source.sortingField).to.be.undefined;
      expect(source.sortDirection).to.be.undefined;
      expect(result.sortingField!.name).to.eq((await provider.getColumns())[0].key);
      expect(result.sortDirection).to.eq(ContentSortDirection.Descending);

      await provider.sort(0, SortDirection.Ascending);
      result = provider.configureContentDescriptor(source);
      expect(source.sortingField).to.be.undefined;
      expect(source.sortDirection).to.be.undefined;
      expect(result.sortingField!.name).to.eq((await provider.getColumns())[0].key);
      expect(result.sortDirection).to.eq(ContentSortDirection.Ascending);

      await provider.sort(0, SortDirection.NoSort);
      result = provider.configureContentDescriptor(source);
      expect(source.sortingField).to.be.undefined;
      expect(source.sortDirection).to.be.undefined;
      expect(result.sortingField!.name).to.eq((await provider.getColumns())[0].key);
      expect(result.sortDirection).to.be.undefined;
    });

    it("sets filterExpression", () => {
      const source = createRandomDescriptor();
      provider.filterExpression = "test";
      const result = provider.configureContentDescriptor(source);
      expect(source.filterExpression).to.be.undefined;
      expect(result.filterExpression).to.eq("test");
    });

  });

  describe("filterExpression", () => {

    it("sets a different filterExpression and clears caches", () => {
      provider.filterExpression = "test 1";
      expect(provider.filterExpression).to.eq("test 1");
      resetMemoizedCacheSpies();

      provider.filterExpression = "test 2";
      expect(provider.filterExpression).to.eq("test 2");
      expect(memoizedCacheSpies.getColumns).to.be.calledOnce;
    });

    it("doesn't clear caches if setting to the same filterExpression", () => {
      provider.filterExpression = "test";
      expect(provider.filterExpression).to.eq("test");
      resetMemoizedCacheSpies();

      provider.filterExpression = "test";
      expect(provider.filterExpression).to.eq("test");
      expect(memoizedCacheSpies.getColumns).to.not.be.called;
    });

  });

  describe("sort", () => {

    it("throws when trying to sort by invalid column", async () => {
      const source = createRandomDescriptor();
      source.fields = [];
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);
      await expect(provider.sort(0, SortDirection.NoSort)).to.eventually.be.rejectedWith(PresentationError);
    });

    it("invalidates descriptor configuration and content", async () => {
      const source = createRandomDescriptor();
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);
      const invalidateCacheMock = moq.Mock.ofInstance(provider.invalidateCache);
      provider.invalidateCache = invalidateCacheMock.object;
      await provider.sort(0, SortDirection.NoSort);
      invalidateCacheMock.verify((x) => x({ descriptorConfiguration: true, content: true }), moq.Times.once());
    });

    it("sets sorting properties", async () => {
      const source = createRandomDescriptor();
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);
      await provider.sort(0, SortDirection.Descending);
      expect(provider.sortColumnKey).to.eq((await provider.getColumns())[0].key);
      expect(provider.sortDirection).to.eq(SortDirection.Descending);
    });

  });

  describe("sortColumn", () => {

    it("returns undefined when no sorting column is set", async () => {
      expect(await provider.sortColumn).to.be.undefined;
    });

    /*it("throws when sorting column key is invalid", async () => {
      const source = createRandomDescriptor();
      presentationManagerMock.setup((x) => x.getContentDescriptor(moq.It.isAny(), moq.It.isAny(), moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);
      await provider.sort(0, SortDirection.Ascending);
      provider.getColumns.cache.clear();
      (provider as any).getDefaultContentDescriptor.cache.clear();
      (provider as any).getContentDescriptor.cache.clear();
      source.fields.splice(0, 1);
      await expect(provider.sortColumn).to.eventually.be.rejectedWith("Assert");
    });*/

    it("returns valid sorting column", async () => {
      const source = createRandomDescriptor();
      presentationManagerMock.setup(async (x) => x.getContentDescriptor({ imodel: imodelMock.object, rulesetId }, moq.It.isAny(), moq.It.isAny(), moq.It.isAny()))
        .returns(async () => source);
      await provider.sort(0, SortDirection.Descending);
      const sortingColumn = await provider.sortColumn;
      expect(sortingColumn).to.eq((await provider.getColumns())[0]);
    });

  });

  describe("getColumns", () => {

    it("returns valid column descriptions", async () => {
      const descriptor = createRandomDescriptor();
      (provider as any).getContentDescriptor = () => descriptor;
      const cols = await provider.getColumns();
      expect(cols).to.matchSnapshot();
    });

    it("returns empty list when descriptor is undefined", async () => {
      (provider as any).getContentDescriptor = () => undefined;
      const cols = await provider.getColumns();
      expect(cols).to.deep.eq([]);
    });

    it("memoizes result", async () => {
      const descriptor = createRandomDescriptor();
      const resultPromiseContainer = new PromiseContainer<Descriptor>();
      const getContentDescriptorMock = moq.Mock.ofInstance(() => (provider as any).getContentDescriptor);
      getContentDescriptorMock.setup((x) => x()).returns(async () => resultPromiseContainer.promise).verifiable(moq.Times.once());
      getContentDescriptorMock.setup((x) => x()).verifiable(moq.Times.never());
      (provider as any).getContentDescriptor = getContentDescriptorMock.object;

      const requests = [1, 2].map(async () => provider.getColumns());
      resultPromiseContainer.resolve(descriptor);
      const response = await Promise.all(requests);
      expect(response[0]).to.deep.eq(response[1], "both responses should be equal");
      presentationManagerMock.verifyAll();
    });

  });

  describe("getRowsCount", () => {

    it("returns count from base class", async () => {
      const size = faker.random.number();
      (provider as any).getContentSetSize = () => size;
      const count = await provider.getRowsCount();
      expect(count).to.eq(size);
    });

  });

  describe("getRow", () => {

    it("returns undefined when content is undefined", async () => {
      (provider as any).getContent = async () => undefined;
      const row = await provider.getRow(0);
      expect(row).to.be.undefined;
    });

    it("returns undefined when content contains no records", async () => {
      (provider as any).getContent = async () => createContent(0);
      const row = await provider.getRow(0);
      expect(row).to.be.undefined;
    });

    it("throws when content record is invalid - contains invalid number of primary keys", async () => {
      const record = createEmptyContentItem();
      record.primaryKeys = [];
      (provider as any).getContent = async (): Promise<Content> => ({
        descriptor: createRandomDescriptor(),
        contentSet: [record],
      });
      await expect(provider.getRow(0)).to.eventually.be.rejectedWith(PresentationError);
    });

    it("requests content in pages", async () => {
      provider = new Provider({
        imodel: imodelMock.object,
        ruleset: rulesetId,
        pageSize: 2,
        cachedPagesCount: 10,
      });
      const contentResolver = [0, 1].map(() => new PromiseContainer<Content>());

      const getContentMock = moq.Mock.ofInstance((provider as any).getContent);
      (provider as any).getContent = getContentMock.object;
      getContentMock.setup((x) => x({ start: 0, size: 2 })).returns(async () => contentResolver[0].promise).verifiable(moq.Times.once());
      getContentMock.setup((x) => x({ start: 2, size: 2 })).returns(async () => contentResolver[1].promise).verifiable(moq.Times.once());

      // request rows without await to make sure paging is handled properly (new
      // pages are not created while other pages for the same position are being loaded)
      const requests = [0, 1, 2].map(async (index) => provider.getRow(index));
      contentResolver.forEach((resolver) => resolver.resolve(createContent(2)));
      const rows = await Promise.all(requests);
      rows.forEach((row) => expect(row).to.not.be.undefined);

      // verify the last row in second page is also loaded (even though we didn't request it)
      expect(provider.getLoadedRow(3)).to.not.be.undefined;

      // verify getContent was called once per page
      getContentMock.verifyAll();
    });

    it("returns valid row item", async () => {
      const descriptor = createRandomDescriptor();
      const values: ValuesDictionary<any> = {};
      const displayValues: ValuesDictionary<any> = {};
      descriptor.fields.forEach((field) => {
        values[field.name] = faker.random.word();
        displayValues[field.name] = faker.random.words();
      });
      const record = new Item([createRandomECInstanceKey()],
        faker.random.words(), faker.random.word(), undefined, values, displayValues, []);
      const c: Content = {
        descriptor,
        contentSet: [record],
      };
      (provider as any).getContent = async () => c;
      const row = await provider.getRow(0);
      expect(row).to.matchSnapshot();
    });

  });

});
