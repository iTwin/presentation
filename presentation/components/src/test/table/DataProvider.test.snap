// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDataProvider getColumns returns one column descriptor when display type is list 1`] = `
Array [
  Object {
    "editable": false,
    "filterable": false,
    "key": "/DisplayLabel/",
    "label": "general.display-label",
    "sortable": true,
  },
]
`;

exports[`TableDataProvider getColumns returns valid column descriptions 1`] = `
Array [
  Object {
    "editable": true,
    "filterable": false,
    "key": "Engineer",
    "label": "monitor",
    "sortable": true,
  },
  Object {
    "editable": false,
    "filterable": false,
    "key": "Home Loan Account",
    "label": "Refined rich",
    "sortable": true,
  },
  Object {
    "editable": true,
    "filterable": false,
    "key": "executive",
    "label": "Director Chicken",
    "sortable": true,
  },
]
`;

exports[`TableDataProvider getRow returns valid row item 1`] = `
Object {
  "cells": Array [
    Object {
      "key": "Home Loan Account",
      "record": PropertyRecord {
        "description": "Books XSS Fresh",
        "property": Object {
          "displayLabel": "Steel",
          "name": "Home Loan Account",
          "typename": "varchar",
        },
        "value": Object {
          "displayValue": "Books XSS Fresh",
          "value": "Automotive",
          "valueFormat": 0,
        },
      },
    },
    Object {
      "key": "Legacy",
      "record": PropertyRecord {
        "description": "synthesizing pixel Avon",
        "property": Object {
          "displayLabel": "Indian Rupee International",
          "editor": Object {
            "name": "Tools",
            "params": Array [],
          },
          "name": "Legacy",
          "typename": "datetime",
        },
        "value": Object {
          "displayValue": "synthesizing pixel Avon",
          "value": "Metrics",
          "valueFormat": 0,
        },
      },
    },
    Object {
      "key": "Credit Card Account",
      "record": PropertyRecord {
        "description": "back up",
        "isReadonly": true,
        "property": Object {
          "displayLabel": "Awesome Wooden Salad regional payment",
          "name": "Credit Card Account",
          "typename": "binary",
        },
        "value": Object {
          "displayValue": "back up",
          "value": "Concrete",
          "valueFormat": 0,
        },
      },
    },
  ],
  "key": "{\\"className\\":\\"full-range\\",\\"id\\":\\"0x163900000001e5c\\"}",
}
`;

exports[`TableDataProvider getRow returns valid row when display type is list 1`] = `
Object {
  "cells": Array [
    Object {
      "key": "/DisplayLabel/",
      "record": PropertyRecord {
        "property": Object {
          "displayLabel": "Label",
          "name": "content_item_label",
          "typename": "string",
        },
        "value": Object {
          "displayValue": "workforce Frozen",
          "value": "workforce Frozen",
          "valueFormat": 0,
        },
      },
    },
  ],
  "key": "{\\"className\\":\\"Home Loan Account\\",\\"id\\":\\"0xe2ed00000039f5\\"}",
}
`;
